scripts.js

```diff
@@ -314,6 +314,198 @@ f.instMeth;
 var b = new Bar();
 b.diffMeth;
 
+// STACKS
+
+// #1
+function Stack() {
+  this.dataStore = [];
+  this.top = 0;
+  this.push = push;
+  this.pop = pop;
+  this.peek = peek;
+  this.clear = clear;
+  this.length = length;
+}
+function push(element) {
+  this.dataStore[this.top++] = element;
+}
+
+function peek() {
+  return this.dataStore[this.top-1];
+}
+
+function pop() {
+  return this.dataStore[--this.top];
+}
+
+function clear() {
+  this.top = 0;
+}
+
+function length() {
+  return this.top;
+}
+// TODO:
+// 1. if a closing prn is missing show the position of the openning prn
+// 2. If there two many closing prns, throw an error
+// 3. Show Bryce the state of a stack for each iteration of a loop
+// 4. Refactor the counter's out(dataStore.length)
+
+function balancedBrackets(string) {
+  var s = new Stack();
+  var counterO = 0;
+  var counterC = 0;
+
+  for (var i = 0; i < string.length; ++i) {
+    if (string[i] == "(") {
+      counterO++;
+      s.push(i);
+    }
+
+    if (string[i] == ")") {
+      counterC++;
+      s.pop(i);
+    }
+  }
+
+   if (counterC > counterO) {
+     console.log("you're missing open prn")
+   } else if (counterC < counterO) {
+
+   } else {
+     console.log('nothing is missing');
+   }
 
+}
+balancedBrackets("(6+4*31+(120/3");
+  
+// #2
+
+function Stack() {
+  this.dataStore = [];
+  this.top = 0;
+  this.push = push;
+  this.pop = pop;
+  this.peek = peek;
+  this.clear = clear;
+  this.length = length;
+}
+function push(element) {
+  this.dataStore[this.top++] = element;
+}
+
+function peek() {
+  return this.dataStore[this.top-1];
+}
+
+function pop() {
+  return this.dataStore.shift();
+}
+
+function clear() {
+  this.top = 0;
+}
+
+function length() {
+  return this.top;
+}
 
+function hasHigherPrecedence(operand) {
+  if((operand == "*") || (operand == "/")) {
+    return true;
+  } else if ((operand == "-") || (operand == "+")){
+    return false;
+  } else {
+    return true;
+  }
+}
+
+function converter(expression) {
+  var s = new Stack;
+  var string = "";
+
+  for (var i = 0; i < expression.length; ++i) {
+    if (!(expression[i].match(/["*:<>?\\/|+\-\[\]]/))) {
+      string += expression[i]
+    }else if (expression[i].match(/["*:<>?\\/|+\-\[\]]/)){
+      if (s.length() != 0 && hasHigherPrecedence(s.peek())) {
```

Comment:
Here I'm checking if the stack is not empty and the first element(operator) in the stack has a higher precedence if yes then I add this operator to the string.